pyasdf/schema.py

```diff
@@ -198,61 +201,57 @@ def validate_remove_default(validator, properties, instance, schema):
-
-    # We can't validate anything that looks like an external
-    # reference, since we don't have the actual content, so we
-    # just have to defer it for now.  If the user cares about
-    # complete validation, they can call
-    # `AsdfFile.resolve_references`.
-    def iter_errors(self, instance, _schema=None, _seen=set()):
-        if id(instance) in _seen:
-            return
-
-        if _schema is None:
-            schema = self.schema
-        else:
-            schema = _schema
-
-        if ((isinstance(instance, dict) and '$ref' in instance) or
-            isinstance(instance, reference.Reference)):
-            return
-
-        if _schema is None:
-            tag = getattr(instance, '_tag', None)
-            if tag is not None:
-                schema_path = self.ctx.tag_to_schema_resolver(tag)
-                if schema_path != tag:
-                    s = load_schema(schema_path, self.ctx.url_mapping)
-                    if s:
-                        with self.resolver.in_scope(schema_path):
-                            for x in self.orig_iter_errors(instance, s):
-                                yield x
-
-            if isinstance(instance, dict):
-                new_seen = _seen | set([id(instance)])
-                for val in six.itervalues(instance):
-                    for x in self.iter_errors(val, _seen=new_seen):
-                        yield x
-
-            elif isinstance(instance, list):
-                new_seen = _seen | set([id(instance)])
-                for val in instance:
-                    for x in self.iter_errors(val, _seen=new_seen):
-                        yield x
-        else:
-            for x in self.orig_iter_errors(instance, _schema=schema):
-                yield x
+    meta_schema = load_schema(YAML_SCHEMA_METASCHEMA_ID,
+                              mresolver.default_url_mapping)
+
+    class ASDFValidator(mvalidators.create(meta_schema=meta_schema,
+                                           validators=validators)):
+        def iter_errors(self, instance, _schema=None, _seen=set()):
+            # We can't validate anything that looks like an external reference,
+            # since we don't have the actual content, so we just have to defer
+            # it for now.  If the user cares about complete validation, they
+            # can call `AsdfFile.resolve_references`.
+            if id(instance) in _seen:
+                return
+
+            if _schema is None:
+                schema = self.schema
+            else:
+                schema = _schema
+
+            if ((isinstance(instance, dict) and '$ref' in instance) or
+                    isinstance(instance, reference.Reference)):
+                return
+
+            if _schema is None:
+                tag = getattr(instance, '_tag', None)
+                if tag is not None:
+                    schema_path = self.ctx.tag_to_schema_resolver(tag)
+                    if schema_path != tag:
```

Comment:
It sort of comes from the fact that there is a generic "Resolver" class whose only purpose is to pattern match strings and convert them into other strings.  (The same mechanism is used to map remote URLs to schema files that ship with pyasdf and thus are already on the local disk).  That could be changed, I suppose, but the resolver thing is really flexible...